libs/storage/Tsavorite/cs/src/core/Index/Recovery/Recovery.cs

```diff
@@ -637,19 +686,27 @@ private async ValueTask RecoverHybridLogAsync(long scanFromAddress, long recover
                 return;
             var recoveryStatus = GetPageRangesToRead(scanFromAddress, untilAddress, checkpointType, out long startPage, out long endPage, out int capacity, out int numPagesToReadFirst);
 
-            // Issue request to read pages as much as possible
-            hlog.AsyncReadPagesFromDevice(startPage, numPagesToReadFirst, untilAddress, hlog.AsyncReadPagesCallbackForRecovery, recoveryStatus);
-
-            for (long page = startPage; page < endPage; page++)
+            for (long page = startPage; page < endPage; page += numPagesToReadFirst)
             {
-                // Ensure page has been read into memory
-                int pageIndex = hlog.GetPageIndexForPage(page);
-                await recoveryStatus.WaitReadAsync(pageIndex, cancellationToken).ConfigureAwait(false);
+                var end = Math.Min(page + numPagesToReadFirst, endPage);
+                var numPagesToReadThisIteration = (int)(end - page);
+
+                // Ensure that page slots that will be read into, have been flushed from previous reads. 
+                await WaitUntilAllPagesHaveBeenFlushedAsync(page, end, recoveryStatus, cancellationToken).ConfigureAwait(false);
```

Comment:
add comment:
// Due to the use of a single read semaphore, this must be done in batches of "all flushes' followed by "all reads" to ensure proper sequencing of reads when usableCapacity != capacity (and thus the page-read index is not equal to the page-flush index).